### PR TITLE
add air purifier

### LINKED ISSUE #2: Are there any plans to support air purifiers?
Hi, thanks for providing a useful sdk.

cocoro also includes an air purifier, and it would be great if you could include an air purifier in the name of Git, is that too much work?

It looks like the TEMPERATURE property names are grouped under "F1" instead of "FA", so I'll have to check the binary again.

### PR DESCRIPTION
**Pull request by shossk**
Closes #2

### CODE REVIEW

File: src/devices/purifier/class.ts [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+			statusCode: StatusCode.POWER,
+			valueType: ValueType.SINGLE,
+		});
+	}
+
+	/**
+	 * Convert mode string to binary.
+	 * @return     {string}  mode binary
+	 */
+	existMode = (mode: string): string => {
```
Comment (dvcrn) [inline src/devices/purifier/class.ts]: existMode -> stringToMode or something like this would be better

File: src/devices/purifier/class.ts [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+	};
+
+	/**
+	 * Queues a specific propertyStatus for change
+	 * This alone does not do anything, the changes need to get transmitted to the
+	 * cocoro air api
+	 *
+	 * @param      {PropertyStatus}  propertyStatus  The property status
+	 */
+	queuePropertyStatusUpdate(propertyStatus: PropertyStatus): void {
```
Comment (dvcrn) [inline src/devices/purifier/class.ts]: This looks identical to device.ts, so we can remove it here. Use the `super.queuePropertyStatusUpdate`

File: src/devices/purifier/properties.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	HIGH = '16',
+}
+
+export enum ValueType {
+	SINGLE = 'valueSingle',
+	BINARY = 'valueBinary',
+	RANGE = 'valueRange',
+}
+
+export interface BinaryPropertyStatus {
```
Comment (dvcrn) [inline src/devices/purifier/properties.ts]: For the types that are identical, just use the ones from the existing properties file - https://github.com/dvcrn/cocoro-sdk/blob/master/src/properties.ts#L72, no need to re-define them

**Comment by dvcrn** [comment]
Looking good! I just don't have a way to test it here. I assume you tried all of these settings with your purifier, right?

File: src/devices/purifier/class.ts (hunk 1 above)
Comment (shossk) [reply src/devices/purifier/class.ts]: All right, thanks!

By the way, would you prefer the private function? I don't plan to use it from outside.

File: src/devices/purifier/class.ts (hunk 2 above)
Comment (shossk) [reply src/devices/purifier/class.ts]: There is a reason for this, the type PropertyStatus is slightly different

I didn't understand the extend part of the type, and I thought I would have to change the device.ts property file to make it the same as device.ts, so I temporarily wrote it separately.

I thought that OPERATION_MODE was unique and included only in the air purifier, so I wanted to separate it, is there any way to solve this?

```typescript
export enum StatusCode {
  STATE_DETAIL = 'FA',
  WINDSPEED = 'A0',
  POWER = '80',
  NANOE_MODE = 'CF',
  OPERATION_MODE = 'B0',
  HEAT_TEMPERATURE = 'B3',
  ROOM_TEMPERATURE = 'BB',
}
```

↓

```typescript
export enum StatusCode {
  STATE_DETAIL = 'FA',
  WINDSPEED = 'A0',
  POWER = '80',
  NANOE_MODE = 'CF',
  OPERATION_MODE = 'B0',
  HEAT_TEMPERATURE = 'B3',
  ROOM_TEMPERATURE = 'BB',
  OPERATION_MODE = 'F3', // ← added this
}
```

File: src/devices/purifier/class.ts (hunk 1 above)
Comment (dvcrn) [reply src/devices/purifier/class.ts]: sure, lets make it private

Comment (shossk) [reply src/devices/purifier/class.ts]: ok! i'm fix it.

File: src/devices/purifier/class.ts (hunk 2 above)
Comment (shossk) [reply src/devices/purifier/class.ts]: I'm sorry, but the solution to this was to make it super and make it as common as possible.
I've made it any, and I know that's not a good solution, so best practices, please!

File: src/devices/purifier/properties.ts (hunk 3 above)
Comment (shossk) [reply src/devices/purifier/properties.ts]: fix it!

**Comment by shossk** [comment]
>Looking good! I just don't have a way to test it here. I assume you tried all of these settings with your purifier, right?

Yes! The air purifier in my house is working just fine because of this!

**Comment by dvcrn** [comment]
For the public interface of this sdk, we're currently getting devices like this:

```
const devices = await cocoro.queryDevices();
const device = devices[0];
```

Is it doable to automatically deduct when a device is a purifier or aircon?

Maybe we could add a separate type field for each one

```
console.log(device.kind); // => Device.Purifier | Device.Aircon
```

**Comment by shossk** [comment]
Do you mean you are looking for a property that can distinguish between an air purifier and an air conditioner?

There is a deviceType as part of the boxinfo, and it looks like you're only looking at part of the boxinfo right now.

I'm sorry if I'm wrong...

```JSON
{
  "box": [
    {
      "echonetData": [
        {
          "labelData": {
            "deviceType": "AIR_CLEANER"
          }
        }
      ]
    }
  ]
}
```

**Comment by dvcrn** [comment]
Ah perfect. Okay can you do one last change? Then we can merge this 👍 

In the constructor to device, can you add a new property for`kind`? https://github.com/dvcrn/cocoro-sdk/blob/master/src/cocoro.ts#L122-L134 If deviceType == 'AIR_CLEANER', set it to a new enum type Device.AirCleaner

**Comment by shossk** [comment]
I've finished adding them!

Added an OR condition for the devices array, remove it if it is not needed.
I'm not sure if it's because I added the OR condition, but the completion doesn't work anymore.
I can see the devices, but the Purifier completion does not work.
Do you have any idea why?

**Comment by dvcrn** [comment]
Thanks! Looks good enough to me. I'll fix that on my end!

Thanks a lot for taking the time to add support for this! It's been very helpful!

**Comment by dvcrn** [comment]
@shossk I've created a separate PR to cleanup some of the code and make things more generic. Can you give it a try to see if your device still works? https://github.com/dvcrn/cocoro-sdk/pull/4

Regarding autocomplete, you need to type-assert the device as Purifier, so (device as Purifier)

**Comment by shossk** [comment]
@dvcrn 
Thanks for the cleanups, they work perfectly!

**Pull request merged**